faiss_vector_posting.go

```diff
@@ -382,179 +382,173 @@ func (sb *SegmentBase) InterpretVectorIndex(field string, requiresFiltering bool
-
-						selector, err = faiss.NewIDSelectorBatch(vectorIDsToInclude)
 					}
-					if err != nil {
-						return nil, err
+
+					selector, err = faiss.NewIDSelectorNot(ineligibleVectorIDs)
+				} else {
+					// Getting the vector IDs corresponding to the eligible doc IDs.
+					// The docVecIDMap maps each docID to vectorIDs corresponding to it.
+					// Usually, each docID has one vecID mapped to it unless
+					// the vector is nested, in which case there can be multiple
+					// vectorIDs mapped to the same docID.
+					// Eg. docID d1 -> vecID v1, for the first case
+					// d1 -> {v1,v2}, for the second case.
+
+					eligibleVecIDsBitmap := roaring.NewBitmap()
+					for eligibleDocIDsIter := eligibleDocIDs.Iterator(); eligibleDocIDsIter.HasNext(); {
+						eligibleVecIDsBitmap.AddMany(docVecIDMap[eligibleDocIDsIter.Next()])
 					}
 
-					// Ordering the retrieved centroid IDs by increasing order
-					// of distance i.e. decreasing order of proximity to query vector.
-					closestCentroidIDs, centroidDistances, _ :=
-						vecIndex.ObtainClustersWithDistancesFromIVFIndex(qVector,
-							eligibleCentroidIDs)
-
-					// Getting the nprobe value set at index time.
-					nprobe := vecIndex.GetNProbe()
-
-					eligibleDocsTillNow := int64(0)
-					minEligibleCentroids := 0
-					for i, centroidID := range closestCentroidIDs {
-						eligibleDocsTillNow += int64(centroidVecIDMap[centroidID].GetCardinality())
-						if eligibleDocsTillNow >= k && i >= int(nprobe-1) {
-							// Continue till at least 'K' cumulative vectors are
-							// collected or 'nprobe' clusters are examined, whichever
-							// comes later.
-							minEligibleCentroids = i + 1
-							break
+					for centroidID, vecIDs := range centroidVecIDMap {
+						vecIDs.And(eligibleVecIDsBitmap)
+						if !vecIDs.IsEmpty() {
+							// The mapping is now reduced to those vectors which
+							// are also eligible docs for the filter query.
+							centroidVecIDMap[centroidID] = vecIDs
+							eligibleCentroidIDs = append(eligibleCentroidIDs, centroidID)
+						} else {
+							// don't consider clusters with no eligible IDs.
+							delete(centroidVecIDMap, centroidID)
 						}
-						minEligibleCentroids = i + 1
 					}
 
-					// Search the clusters specified by 'closestCentroidIDs' for
-					// vectors whose IDs are present in 'vectorIDsToInclude'
-					scores, ids, err := vecIndex.SearchClustersFromIVFIndex(
-						selector, len(vectorIDsToInclude), closestCentroidIDs,
-						minEligibleCentroids, k, qVector, centroidDistances, params)
-					if err != nil {
-						return nil, err
+					selector, err = faiss.NewIDSelectorBatch(vectorIDsToInclude)
+				}
+				if err != nil {
+					return nil, err
+				}
+
+				// Ordering the retrieved centroid IDs by increasing order
+				// of distance i.e. decreasing order of proximity to query vector.
+				closestCentroidIDs, centroidDistances, _ :=
+					vecIndex.ObtainClustersWithDistancesFromIVFIndex(qVector, eligibleCentroidIDs)
```

Comment:
@metonymic-smokey I'm hoping profiling will reveal how much of time/CPU is distributed among ..
- ObtainClustersWithDistancesFromIVFIndex
- SearchClustersFromIVFIndex